comp/logs/agent/agent.go

```diff
@@ -0,0 +1,83 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package agent
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"go.uber.org/fx"
+
+	"github.com/djmitche/dd-agent-comp-experiments/comp/core/config"
+	"github.com/djmitche/dd-agent-comp-experiments/comp/core/log"
+	"github.com/djmitche/dd-agent-comp-experiments/comp/core/status"
+	"github.com/djmitche/dd-agent-comp-experiments/comp/logs/internal"
+	"github.com/djmitche/dd-agent-comp-experiments/comp/logs/launchers/launchermgr"
+)
+
+type agent struct {
+	log         log.Component
+	launchermgr launchermgr.Component
+}
+
+type dependencies struct {
+	fx.In
+
+	Lc          fx.Lifecycle
+	Params      internal.BundleParams
+	Config      config.Component
+	Log         log.Component
+	Launchermgr launchermgr.Component
+}
+
+type provides struct {
+	fx.Out
+
+	Component
+	StatusReg *status.Registration `group:"true"`
+}
+
+func newAgent(deps dependencies) provides {
+	a := &agent{
+		log:         deps.Log,
+		launchermgr: deps.Launchermgr,
+	}
+
+	prov := provides{Component: a}
+	if deps.Params.ShouldStart(deps.Config) {
```

Comment:
Is the line `if deps.Params.ShouldStart(deps.Config)` needed for each sub component of the log agent?

Reply:
It is, for the same reasons I mentioned above -- I think it's important to allow instantiating components without starting them, since instantiation is easy to do "accidentally" and it's difficult to see what the implications might be.

Reply:
If all components for the log agent must have an "enable logic", it will requires lot of typing but also there will be a lot of room for mistakes. I am also pretty sure there are other components that are enable by a config flag.

When looking at the code, I find that https://github.com/DataDog/dd-agent-comp-experiments/pull/1#discussion_r934665971 is probably relevant to avoid the previous issues. What do you think?

Reply:
I _really_ want to avoid nesting fx.App's.  I see that that might work if we're careful to fx.Supply everything from the parent App to the child App, but it seems like that will be difficult and I'm sure there are other demons lurking.  So that leaves the first option in that comment:

---
1. The config can be instantiate outside `fx`  and `fx.Invoke(func( agent logsagent.Component)` can be added conditionally based on `config.GetBool("logs_enabled")`
```go
        var options []fx.Option
	options = append(options, fx.Provide(newLogsAgent))
	options = append(options, fx.Invoke(Run)) // Other run function
	if config.GetBool("logs_enabled") {
		options = append(options, fx.Invoke(RunLogAgent))
	}
	app := fx.New(options...)
``` 
---
which is pretty similar to what I've done here.  The differences are:
 * Locates the check of the config parameter within the component itself, rather than across several `cmd/whatever/main.go`'s
 * Locates the check in the constructor (newLogsAgent in this case) rather than in a function that runs after the constructor (RunLogAgent)
 * Does not require that config be available before app startup

What's not different is
 * The enable signal has to be transmitted to all of the active components in logs-agent, somehow

I agree that it's a bit ugly that every subcomponent must check this value, and especially each component must then require `comp/core/config` and pass it to `ShouldStart(config)`.  But, whatever we do, each of those components has to decide, "should I start?" so some amount of code is required.  `if blahblahblah.ShouldStart(blah) { .. }` seems about as small as we could get that.

The second difference is important: because we've removed the significance of the "setup phase", just about everything about component relationships needs to be known in the constructor.  In particular, event subscriptions need to be made (or not) in the constructor.